File: frontend/projects/irrtirer/src/app/core/helpers/polygon/point-to-polygon-distance-calculator.ts
import { Vector } from '../../models/vector.model';

export class PointToPolygonDistanceCalculator {
    public static calculatePointToPolygonDistance(point: Vector, vertices: Vector[]): number {
        let minDistance = Number.MAX_VALUE;

        for (let i = 0; i < vertices.length; i++) {
            const p1: Vector = vertices[i];
            const p2: Vector = vertices[(i + 1) % vertices.length];

            const r = Vector.dot(p2.sub(p1), point.sub(p1)) / Math.pow(p2.sub(p1).magnitude(), 2);

            let distance;
            if (r < 0) {
                distance = point.sub(p1).magnitude();
            } else if (r > 1) {
                distance = p2.sub(point).magnitude();
            } else {
                distance = Math.sqrt(Math.pow(point.sub(p1).magnitude(), 2) - Math.pow(r * p2.sub(p1).magnitude(), 2));
            }

            if (distance < minDistance) {
                minDistance = distance;
            }
        }

        return minDistance;
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Helpers/PolygonHelpers/PointToPolygonDistanceCalculator.cs
using System;
using System.Numerics;

namespace Irrtirer.Generator.Helpers.PolygonHelpers
{
    internal static class PointToPolygonDistanceCalculator
    {
        public static float CalculatePointToPolygonDistance(Vector2 point, Vector2[] vertices)
        {
            float minDistance = float.MaxValue;

            for (int i = 0; i < vertices.Length; i++)
            {
                Vector2 p1 = vertices[i];
                Vector2 p2 = vertices[(i + 1) % vertices.Length];

                var r = Vector2.Dot(p2 - p1, point - p1) / MathF.Pow((p2 - p1).Length(), 2);

                float distance;
                if (r < 0)
                {
                    distance = (point - p1).Length();
                }
                else if (r > 1)
                {
                    distance = (p2 - point).Length();
                }
                else
                {
                    distance = MathF.Sqrt(MathF.Pow((point - p1).Length(), 2) - MathF.Pow(r * (p2 - p1).Length(), 2));
                }

                if (distance < minDistance)
                {
                    minDistance = distance;
                }
            }

            return minDistance;
        }
    }
}
